### PR TITLE
Windows install - use an app ID with toaster popups

### PR DESCRIPTION
**Pull request by SimonWalton-HiFi**
https://highfidelity.manuscript.com/f/cases/15094/Fix-SnoreToast-application-name-showing-up-for-Sandbox-via-Electron-update-notifications

The update notification popup for Win uses the SnoreToast application. This requires a valid appUserModelId to put the app name in the popup instead of "SnoreToast".

This PR adds a canned name for the installer to assign (the Windows auto-generated one looks fragile).
It requires a new plug-in for NSIS.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10084/HighFidelity-Beta-PR13230-f0570557b24b67aba4906da02104e76ed60fcefe.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10084/HighFidelity-Beta-PR13230-f0570557b24b67aba4906da02104e76ed60fcefe.dmg

**Comment by SimonWalton-HiFi** [comment]
The required plug-in for the Win build machine is
ApplicationID:

http://nsis.sourceforge.net/ApplicationID_plug-in

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10129/HighFidelity-Beta-PR13230-65bb4c218ebfcba766391308f36acc3b01a8d8f1.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10129/HighFidelity-Beta-PR13230-65bb4c218ebfcba766391308f36acc3b01a8d8f1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10129/HighFidelity-Beta-Interface-PR13230-65bb4c218ebfcba766391308f36acc3b01a8d8f1.dmg

**Comment by SimonWalton-HiFi** [comment]
I haven't been able to get it to work with electron's setAppUserModelId(). SnoreToast might be fussy about where the ID comes from.

I can probably put the ID on the Sandbox shortcut to get the Sandbox label in the popup.

**Comment by birarda** [comment]
@SimonWalton-HiFi sounds good - though I think we should do it in both places

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10219/HighFidelity-Beta-PR13230-ef4d09c.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10219/HighFidelity-Beta-PR13230-ef4d09c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10219/HighFidelity-Beta-Interface-PR13230-ef4d09c.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10234/HighFidelity-Beta-PR13230-5e17ef5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10234/HighFidelity-Beta-PR13230-5e17ef5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10234/HighFidelity-Beta-Interface-PR13230-5e17ef5.dmg

File: cmake/templates/NSIS.template.in [hunk 1]
```diff
@@ -1173,6 +1173,18 @@ Section "-Core installation"
...
+    ${If} @SERVER_COMPONENT_CONDITIONAL@
+      ApplicationID::Set "$SMPROGRAMS\$STARTMENU_FOLDER\@CONSOLE_SHORTCUT_NAME@.lnk" "com.highfidelity.sandbox"
+    ${EndIf}
+    ApplicationID::Set "$SMPROGRAMS\$STARTMENU_FOLDER\@INTERFACE_SHORTCUT_NAME@.lnk" "com.highfidelity.interface"
+  !else
+    ${If} @SERVER_COMPONENT_CONDITIONAL@
+      ApplicationID::Set "$SMPROGRAMS\$STARTMENU_FOLDER\@CONSOLE_SHORTCUT_NAME@.lnk" "com.highfidelity.sandbox-dev"
+    ${EndIf}
+    ApplicationID::Set "$SMPROGRAMS\$STARTMENU_FOLDER\@INTERFACE_SHORTCUT_NAME@.lnk" "com.highfidelity.interface-dev"
+  !endif
```
[the remark this answers is not in the document]
Comment (SimonWalton-HiFi) [reply cmake/templates/NSIS.template.in]: Done.

File: server-console/src/main.js [hunk 2]
```diff
@@ -44,6 +44,8 @@ const LOG_FILE_REGEX = /(domain-server|ac-monitor|ac)-.*-std(out|err).txt/;
 
 const HOME_CONTENT_URL = "http://cdn.highfidelity.com/content-sets/home-tutorial-RC40.tar.gz";
 
+const APP_USER_MODEL_ID = 'com.highfidelity.sandbox';  // For Windows jump list, etc.
```
[the remark this answers is not in the document]
Comment (SimonWalton-HiFi) [reply server-console/src/main.js]: Done.

**Comment by SimonWalton-HiFi** [comment]
I'm now passing down the app user-model ID from CMake to build-info.json. I'm not sure where this file gets built in production; development packages don't seem to create it.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10323/HighFidelity-Beta-PR13230-97613f1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10323/HighFidelity-Beta-Interface-PR13230-97613f1.dmg

**Comment by SimonWalton-HiFi** [comment]
## Test Plan

Build a production release installer for client & server (I'm not sure exactly how one does this).

For Windows:
Install the client & server normally. Start the sandbox if it doesn't automatically start. If there is a later release available a windows notification should appear. To force the sandbox to think a later version is available, if necessary, edit /Program Files/High Fidelity/server-console/build-info.json and change buildIdentifier to "1".

With this PR the popup should say "An update is available! ...". There should be no SnoreToast label - see linked Manuscript for details.

Other platforms:
Should be no change.

**Comment by birarda** [comment]
Approved, QA awaiting ops installation of NSIS plugin

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10557/HighFidelity-Beta-PR13230-01ef636.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10564/HighFidelity-Beta-PR13230-66cad78.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10564/HighFidelity-Beta-PR13230-66cad78.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10564/HighFidelity-Beta-Interface-PR13230-66cad78.dmg

**Comment by SimonWalton-HiFi** [comment]
With the latest merged installer changes I have to set buildNumber to "1" and stableBuild to "0" to get the popup. The pop label is "Sandbox -"; I think this is because BUILD_VERSION_NO_SHA isn't defined for my installer. The production build label should just be "Sandbox".

**Comment by SimonWalton-HiFi** [comment]
The installer should now be adding the AppUserModelId to the desktop shortcut for Sandbox (if their installation is selected) also. Tested starting Sandbox from desktop and start menu - popup appeared correctly in all cases.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10570/HighFidelity-Beta-PR13230-ce7a10c.apk

**Comment by SimonWalton-HiFi** [comment]
## Test Plan Supplement
Ensure the popup functions correctly whether the Sandbox is initially started from the installer, desktop shortcut, start menu or via Windows explorer (etc).

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10570/HighFidelity-Beta-PR13230-ce7a10c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10570/HighFidelity-Beta-Interface-PR13230-ce7a10c.dmg

**Comment by birarda** [comment]
### Test Plan

Run Sandbox from this PR. Confirm that it does not prompt you to update to a later version.

For each of the two following builds:
- http://hifi-public.s3.amazonaws.com/birarda/installers/HighFidelity-Beta-0.68.0.exe
- http://hifi-public.s3.amazonaws.com/birarda/installers/HighFidelity-Beta-8300-ce7a10c.exe

Install the build and chose the option to add a desktop shortcut. 
Run the Sandbox using the desktop shortcut and confirm it prompts you to update. 
Confirm that the application name displayed in the notification is "Sandbox" and not "SnoreToast" or "com.highfidelity.sandbox".
Quit the Sandbox and run it again from the Start Menu. 
Confirm that the application name displayed in the notification is "Sandbox" and not "SnoreToast" or "com.highfidelity.sandbox".
Quit the Sandbox and run it directly from the install folder. 
Confirm that the application name displayed in the notification is "Sandbox" and not "SnoreToast" or "com.highfidelity.sandbox".

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10580/HighFidelity-Beta-PR13230-ce7a10c.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10580/HighFidelity-Beta-PR13230-ce7a10c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10580/HighFidelity-Beta-Interface-PR13230-ce7a10c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10580/HighFidelity-Beta-PR13230-ce7a10c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10580/HighFidelity-Beta-Interface-PR13230-ce7a10c.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10582/HighFidelity-Beta-PR13230-4807f9d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10582/HighFidelity-Beta-PR13230-4807f9d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10582/HighFidelity-Beta-Interface-PR13230-4807f9d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10582/HighFidelity-Beta-PR13230-4807f9d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10582/HighFidelity-Beta-Interface-PR13230-4807f9d.exe

**Comment by SimonWalton-HiFi** [comment]
I tried the full Win installer (HighFidelity-Beta-PR13230-4807f9d) above. The first couple of times I had another domain server running and it seemed like the installer exited without a warning popup about an existing DS. After stopping the DS the installer ran correctly. I edited build-info.json and started from the desktop link. The new-build popup worked correctly with the right label. Uninstall worked correctly.

**Comment by qliejahner** [comment]
Currently testing

**Comment by qliejahner** [comment]
PASS @conklin94122 

The sandbox from this PR did not create a notification. The linked builds created notifications in all expected scenarios, all of which referred to "Sandbox" as the app name.

**Pull request merged**